R/intervals.R

```diff
@@ -0,0 +1,625 @@
+#'
+#' @rdname problematic_intervals
+identify_missing_intervals <- function(ints_dt,
+                                       full_int_start = 0,
+                                       full_int_end = Inf) {
+
+  assertive::is_data.table(ints_dt)
+  assertthat::is.number(full_int_start)
+  assertthat::is.number(full_int_end)
+
+  # create full interval that all sub intervals should span
+  full_int <- intervals::Intervals_full(
+    matrix(c(full_int_start, full_int_end), ncol = 2), closed = c(TRUE, FALSE)
+  )
+
+  # create left-closed, right-open intervals
+  ints <- intervals::Intervals_full(as.matrix(ints_dt), closed = c(TRUE, FALSE))
+
+  # identify missing intervals
+  missing_ints <- intervals::interval_difference(full_int, ints)
+
+  missing_ints_dt <- data.table::as.data.table(missing_ints)
+  data.table::setnames(missing_ints_dt, c("start", "end"))
+  data.table::setkeyv(missing_ints_dt, c("start", "end"))
+  return(missing_ints_dt)
+}
+
+#' @rdname problematic_intervals
+identify_overlapping_intervals <- function(ints_dt,
+                                           full_int_start = 0,
+                                           full_int_end = Inf) {
+
+  assertive::is_data.table(ints_dt)
+  assertthat::is.number(full_int_start)
+  assertthat::is.number(full_int_end)
+
+  # create left-closed, right-open intervals
+  ints <- intervals::Intervals_full(as.matrix(ints_dt), closed = c(TRUE, FALSE))
+
+  # get list mapping between intervals if they overlap at all
+  overlaps <- intervals::interval_overlap(ints, ints)
+  names(overlaps) <- 1:length(overlaps)
+
+  # sort by number of intervals that each interval overlaps with so that we can
+  # identify the largest overlapping intervals first
+  overlaps <- overlaps[order(sapply(overlaps, length), decreasing=T)]
+
+  overlapping_indices <- c()
+  for (i in names(overlaps)) {
+    # remove match to itself
+    overlaps[[i]] <- overlaps[[i]][overlaps[[i]] != i]
+
+    # remove indices of overlapping intervals that have already been identified
+    overlaps[[i]] <- overlaps[[i]][!overlaps[[i]] %in% overlapping_indices]
+
+    if (length(overlaps[[i]]) > 0) {
+      overlapping_indices <- c(overlapping_indices, i)
+    }
+  }
+  overlapping_ints <- ints[as.integer(overlapping_indices)]
+
+  overlapping_ints_dt <- data.table::as.data.table(overlapping_ints)
+  data.table::setnames(overlapping_ints_dt, c("start", "end"))
+  data.table::setkeyv(overlapping_ints_dt, c("start", "end"))
+  return(overlapping_ints_dt)
+}
+
+#' @title create new interval nodes for any missing interval nodess
+#'
+#' @description Check that the children nodes cover the entire parent interval node
+#'   and create new interval nodes for any missing ranges.
+#'
+#' @param interval_tree \[`data.tree()`\]\cr
+#'   interval tree containing the subtree to be modified
+#' @param subtrees \[`list(data.tree())`\]\cr
+#'   non-leaf subtrees to check and fill any missing intervals.
+#' @inheritParams agg
+#'
```

Comment:
Add example

Reply:
I was thinking the package internal utility functions don't need examples

Reply:
An example might be useful as (a) something for people developing/editing these functions after us, (b) a test when you compile the package.

Reply:
It would take some work to come up short examples for some of the helper functions because they are dependent on data from the other functions but I agree it would be helpful for people improving these functions in the future.

I do think we are already doing a better job than usual of documenting non-exported functions. Like a lot of the utility functions in dplyr are not documented at all. https://github.com/tidyverse/dplyr/blob/master/R/utils-format.r

Reply:
They are indirectly tested when testing the exported functions that these utility functions help